File: spark-demo1/src/main/java/com/yangyh/java/code/demo03/secondsort/Demo01SecondSort.java
package com.yangyh.java.code.demo03.secondsort;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import scala.Tuple2;

/**
 * @description:
 * @author: yangyh
 * @create: 2019-12-13 11:02
 */
public class Demo01SecondSort {

    public static void main(String[] args) {

        SparkConf conf = new SparkConf();
        conf.setMaster("local");
        conf.setAppName("secondSort");

        JavaSparkContext sc = new JavaSparkContext(conf);

        JavaRDD<String> lines = sc.textFile("./data/secondSort.txt");

        JavaPairRDD<MySort, String> mySortStringJavaPairRDD = lines.mapToPair(line -> {
            String[] array = line.split(" ");
            Integer first = Integer.valueOf(array[0]);
            Integer second = Integer.valueOf(array[1]);
            return new Tuple2<MySort, String>(new MySort(first, second), line);
        });

        mySortStringJavaPairRDD.sortByKey(false).foreach(tp -> System.out.println(tp._2));

    }
}

File: spark-demo1/src/main/scala/com/yangyh/scala/sparkcore/demo03/secondsort/Demo01SecondSort.scala
package com.yangyh.scala.sparkcore.demo03.secondsort

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}


case class MySort(first: Int, second: Int) extends Ordered[MySort] {
  override def compare(that: MySort): Int = {
    if(this.first == that.first) {
      this.second - that.second
    } else {
      this.first - that.first
    }
  }
}

/**
 * 二次排序
 */
object Demo01SecondSort {

  def main(args: Array[String]): Unit = {
    val conf = new SparkConf()
    conf.setMaster("local")
    conf.setAppName("二次排序")

    val sc = new SparkContext(conf)

    val lines: RDD[String] = sc.textFile("./data/secondSort.txt")

    val rdd1: RDD[(MySort, String)] = lines.map(line => {
      val words = line.split(" ")
      Tuple2(MySort(words(0).toInt, words(1).toInt), line)
    })

    rdd1.sortByKey(false).foreach(elem => println(elem._2))

  }

}
